src/View/Sidebar/DeviceListBox.vala

```diff
@@ -58,38 +48,71 @@ public class Sidebar.DeviceListBox : Gtk.ListBox, Sidebar.SidebarListInterface {
+
+        return strcmp (key1, key2);
+    }
+
+    private AbstractMountableRow add_bookmark (string label, string uri, Icon gicon,
                                     string? uuid = null,
                                     Drive? drive = null,
                                     Volume? volume = null,
                                     Mount? mount = null,
                                     bool pinned = true,
                                     bool permanent = false) {
 
-        DeviceRow? bm = has_uuid (uuid, uri);
-
-        if (bm == null) {
-            var new_bm = new DeviceRow (
-                label,
-                uri,
-                gicon,
-                this,
-                pinned, // Pin all device rows for now
-                permanent || (bm != null && bm.permanent), //Ensure bind mount matches permanence of uuid
-                uuid != null ? uuid : uri, //uuid fallsback to uri
-                drive,
-                volume,
-                mount
-            );
+        AbstractMountableRow? bm = null; // Existing bookmark with same uuid
+        if (!has_uuid (uuid, uri, out bm) || bm.custom_name != label) { //Could be a bind mount with the same uuid
+            AbstractMountableRow new_bm;
+            if (drive != null && volume == null) {
+                new_bm = new DriveRow (
+                    label,
+                    uri,
+                    gicon,
+                    this,
+                    pinned, // Pin all device rows for now
+                    permanent || (bm != null && bm.permanent), //Ensure bind mount matches permanence of uuid
+                    uuid != null ? uuid : uri, //uuid fallsback to uri
+                    drive
+                );
+            } else if (volume == null ) {
+                new_bm = new VolumelessMountRow (
+                    label,
+                    uri,
+                    gicon,
+                    this,
+                    pinned, // Pin all device rows for now
+                    permanent || (bm != null && bm.permanent), //Ensure bind mount matches permanence of uuid
+                    uuid != null ? uuid : uri, //uuid fallsback to uri
+                    mount
+                );
+            } else {
+                new_bm = new VolumeRow (
+                    label,
+                    uri,
+                    gicon,
+                    this,
+                    pinned, // Pin all device rows for now
+                    permanent || (bm != null && bm.permanent), //Ensure bind mount matches permanence of uuid
+                    uuid != null ? uuid : uri, //uuid fallsback to uri
+                    volume
+                );
+            }
 
             add (new_bm);
 
-            return new_bm;
+            show_all ();
+            bm = new_bm;
+            bm.update_free_space ();
         }
 
-        return bm;
+        return bm; // Should not be null (either an existing bookmark or a new one)
```

Comment:
I guess adding an `assert (bm != null);` before the return statement would be good to spot potential flaws in the future?

Reply:
Done